### PR TITLE
[compiler] Infer optional dependencies (behind feature flag)

### PR DESCRIPTION
**Pull request by josephsavona**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #30840
* #30838
* #30836
* __->__ #30819
* #30816
* #30814
* #30813
* #30812
* #30811

Adds a new feature flag, `@enableOptionalDependencies` which when enabled allows PropagateScopeDeps and DeriveMinimalDeps to infer optional dependency paths (`a?.b`). 

In PropagateScopeDeps we look for specific safe patterns of nested optional member expressions:
* `<variable> "." / "?." <property>`
* or `<nested> "." / "?." <property>`

When we find this pattern we record a dependency on the overall chain, so for example in `a?.b.c?.d.map()` we would record a dependency on `a?.b.c?.d` (because the outer `.map()` portion doesn't match the above structure). If the structure doesn't match - for example with `a?.b?.[foo(bar)]?.z` - then we fall back to the existing behavior which treats everything after the initial portion as conditional (for that last example we'd continue to record `a` as the dep).

The other portion is DeriveMinimalDeps, which now represents optional access/dependency states and updates the merge logic to handle them. The order of precedence is unconditional > optional > conditional. 

Note: the flag is off by default, but i tried enabling it for all fixtures and the results were all either unchanged or had correct optional deps inferred. I'll also try running it internally with the flag enabled, but i think we can proceed with review and fix-forward any issues identified from that testing.

### CODE REVIEW

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/optional-member-expression-as-memo-dep.expect.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import { c as _c } from "react/compiler-runtime"; // @validatePreserveExistingMemoizationGuarantees
+function Component(props) {
+  const $ = _c(4);
+
+  props?.items.edges?.nodes;
+  let t0;
+  let t1;
+  if ($[0] !== props?.items.edges?.nodes) {
+    t1 = props?.items.edges?.nodes.map();
+    $[0] = props?.items.edges?.nodes;
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/optional-member-expression-as-memo-dep.expect.md:27]: woohoo!

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/reduce-reactive-deps/memberexpr-join-optional-chain2.expect.md [hunk 2]
```diff
@@ -21,12 +21,20 @@ export const FIXTURE_ENTRYPOINT = {
...
     x.push(props.items?.length);
-    x.push(props.items?.edges?.map?.(render)?.filter?.(Boolean) ?? []);
+    let t0;
+    if ($[2] !== props) {
+      t0 = props.items?.edges?.map?.(render)?.filter?.(Boolean) ?? [];
+      $[2] = props;
+      $[3] = t0;
+    } else {
+      t0 = $[3];
+    }
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/reduce-reactive-deps/memberexpr-join-optional-chain2.expect.md:36]: this is weird

File: compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/PropagateScopeDependencies.ts [hunk 3]
```diff
@@ -744,51 +747,212 @@ class PropagationVisitor extends ReactiveFunctionVisitor<Context> {
...
+        sequence.instructions[0].value.instructions[0].lvalue.identifier.id &&
+      sequence.value.kind === 'SequenceExpression' &&
+      sequence.value.instructions.length === 1 &&
+      sequence.value.instructions[0].lvalue !== null &&
+      sequence.value.instructions[0].value.kind === 'PropertyLoad' &&
+      sequence.value.instructions[0].value.object.identifier.id ===
+        sequence.instructions[0].value.value.place.identifier.id &&
+      sequence.value.value.kind === 'LoadLocal' &&
+      sequence.value.value.place.identifier.id ===
+        sequence.value.instructions[0].lvalue.identifier.id
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/PropagateScopeDependencies.ts:842]: what

Comment (josephsavona) [reply compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/PropagateScopeDependencies.ts:842]: the

File: compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/PropagateScopeDependencies.ts [hunk 4]
```diff
@@ -744,51 +747,212 @@ class PropagationVisitor extends ReactiveFunctionVisitor<Context> {
...
+      suggestions: null,
+    });
+    // Instructions are the unconditionally executed portion before the `?`
+    for (const instr of inner.instructions) {
+      this.visitInstruction(instr, context);
+    }
+    // The final value is the conditional portion following the `?`
+    context.enterConditional(() => {
+      this.visitReactiveValue(context, id, inner.value, lvalue);
+    });
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/PropagateScopeDependencies.ts:927]: same code as before if the pattern doesn't match

File: compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/PropagateScopeDependencies.ts (hunk 3 above)
Comment (josephsavona) [reply compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/PropagateScopeDependencies.ts:842]: This is bonkers but also temporary, we won’t need this once PropagateScopeDeps is HIR-based. So seems fine for now - worst case it doesn’t match and we fall back to the previous behavior.

**Review by mofeiZ — APPROVED** [review]
Makes sense, looking forward to understanding / cleaning up the sequence expression logic

**Pull request merged**